magma/find_unconnected_ports.py

```diff
@@ -0,0 +1,136 @@
+        # automatically wire it downstream.
+        if isinstance(port, ClockTypes):
+            msg = "{} not driven, will attempt to automatically wire"
+            _logger.debug(WiringLog(msg, port), debug_info=debug_info)
+            return
+        # If there is a clock nested in @port, then we need to traverse its
+        # children. Note that although we use is_clock_or_nested_clock here,
+        # because of the preceding check for ClockTypes, this conditional will
+        # only catch the nested clock case.
+        if is_clock_or_nested_clock(type(port)):
+            for elt in port:
+                yield from find_unconnected_ports_of_port(elt, debug_info)
+            return
+        yield UnconnectedPortInfo(port, debug_info)
+        return
+    return
+
+
+def find_unconnected_ports_of_interface(interface, debug_info):
+    for port in interface.ports.values():
+        yield from find_unconnected_ports_of_port(port, debug_info)
+
+
+def find_unconnected_ports_of_circuit(ckt):
+    yield from find_unconnected_ports_of_interface(
+        ckt.interface, ckt.debug_info
+    )
+    for inst in ckt.instances:
+        yield from find_unconnected_ports_of_interface(
+            inst.interface, inst.debug_info
+        )
+
+
+@dataclasses.dataclass
+class UnconnectedPortDiagnostic:
+    value: Type
+    connected: Optional[bool]
+    depth: int
+
+    def make_wiring_log(self) -> str:
+        tab = " " * (4 * self.depth)
+        connected_str = ""
+        if self.connected is True:
+            connected_str = ": Connected"
+        elif self.connected is False:
+            connected_str = ": Unconnected"
+        tpl = f"{tab}{{}}{connected_str}"
+        return WiringLog(tpl, self.value)
+
+
+# NOTE(rsetaluri): The value visitor class for constructing the unconnected port
+# diagnostic info requires inheriting from ValueVisitor (defined in
+# magma/value_utils.py). This would result in a cyclical import dependency. To
+# hack around this issue, the value visitor class is created in a factory which
+# should only run once, and use the cached result for future invocations. Note
+# that we set @maxsize to 1 for the lru cache.
+@functools.lru_cache(maxsize=1)
+def _make_unconnected_port_diagnostic_visitor_cls():
+    from magma.value_utils import ValueVisitor
+
+    class _Visitor(ValueVisitor):
+        def __init__(self):
+            self._depth = 0
+
+        def visit(self, node):
+            yield from super().visit(node)
+
+        def generic_visit(self, value):
+            if value.trace() is not None:
+                yield UnconnectedPortDiagnostic(value, True, self._depth)
+                return
+            wrapped = self.wrap(value)
+            results = []
+            has_connected = False
+            self._depth += 1
+            for child in wrapped.children:
+                for diagnostic in self.visit(child):
+                    has_connected = has_connected or diagnostic.connected
```

Comment:
```suggestion
                    has_connected |= diagnostic.connected
```

Reply:
@leonardt actually this doesn't work since it relies on diagnostic.connected being possibly a NoneType (and that doesn't work for bitwise or).